File: 0050-powx-n/0050-powx-n.cpp
class Solution {
public:
    double myPow(double x, long n) {
        bool powneg = n<0;
        bool neg_val = x<0;
        if(abs(x)==1) return n%2==0?1:neg_val?-1:1;
        if(neg_val) neg_val=abs(n)&1;
        double res=1; 
        x=abs(x);
        n=abs(n);
        while(n>0 && res<=100000){
            if(n&1==0)
            {
                x=(x*x);
                n>>=1;
            }
            else{
                res*=x;
                n=n-1;
            }
        }
        res=powneg?1/res:res;
        return neg_val?-res:res;
    }

    // double powerVal(double x,long n){
    //     if(n==1) return x;
    //     if(n==0) return 1;
    //     if(n%2) return x*powerVal(x,(n-1));
    //     double ans = powerVal(x,n>>1);
    //     return ans*ans;
    // }
};

File: 0050-powx-n/0050-powx-n.py
class Solution:
    def calPow(self,val:float,power:int)->float:
        if power==0: 
            return 1 
        if power==1: 
            return val 
        if power%2==0: 
            ans = self.calPow(val,power/2)
            return ans*ans
        return val*self.calPow(val,power-1)
    
    def myPow(self, x: float, n: int) -> float:
        pow_neg,val_neg=False,False  
        if n<0: 
            pow_neg=True 
            n=abs(n)
        if x<0:  
            if n%2: 
                val_neg = True 
            x=abs(x)
        if x!=1:
            res = self.calPow(x,n)
        else: 
            res = 1 
        res =  -res if val_neg else res 
        res =  1/res if pow_neg else res 
        return res


